### ISSUE TITLE
poisson distribution in edward

### ISSUE THREAD

**Issue by dustinvtran** [blei-lab/edward#403]
from an e-mail exchange with @stephma:

 > ... had difficulties when running simple models like poisson factorization when there is a Poisson distribution involved in Edward — have you ever encountered similar problems? How reliable and robust is Edward in this regard?

**Comment by dustinvtran** [comment]
the poisson distribution doesn’t currently support sampling. this presents a difficulty if you instantiate the `Poisson` random variable in edward. there’s a workaround you can do if you only use poisson as a likelihood (as in poisson MF). to do this, see the model section of http://edwardlib.org/tutorials/latent-space-models, which has a very similar matrix factorization model and also uses poisson likelihood. (also see #396)

another note is whether BBVI for poisson factorization is sufficiently fast, compared to the custom coordinate wise updates in the poisson factorization papers. i don’t think it would scale very well, so you may have to resort to implementing the coordinate wise updates (possibly in edward, or using alternative MF-specific software altogether).

**Comment by shoyer** [comment]
This has been fixed upstream in TensorFlow now! https://github.com/tensorflow/tensorflow/issues/6798

**Comment by dustinvtran** [comment]
this is excellent news! once it's available in a version release, we can try out some of the following models:

+ latent space model
+ poisson process
+ cox process
+ poisson factorization
+ poisson deep exponential families

**Comment by gundeep59** [comment]
Hi, has this been released? I am trying to work out the model mention in [this paper](http://www.cs.columbia.edu/~blei/papers/GopalanHofmanBlei2015.pdf). I am using the gamma distribution for inference, but the loss during inference is shown to be nan. The code, inspired from PMF example is here,
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
import edward as ed
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf
from edward.models import Poisson, Gamma

def buildToyDataset(N,M, noise_std = 0.1):
    R = np.random.randint(2, size=(N, M))
    return R

def get_indicator(N,M,prob=0.5): ## chooses which values are present in the training , again a part of toy creation
    ind = np.random.binomial(1,prob,(N,M))
    return ind

## Global Variables needed for toy dataset
N = 50  ## number of users
M = 60  ## number of movies/items
D = 3   ## number of latent factors
## Construct data
R_true = buildToyDataset(N,M)
I_train = get_indicator(N,M) 
I_test = 1 - I_train

## Priors
a_prior = tf.constant(0.2, dtype=tf.float32)
b_prior = tf.constant(3.,dtype=tf.float32)
c_prior = tf.constant(0.1, dtype=tf.float32)
d_prior = tf.constant(4., dtype=tf.float32)
a = tf.constant(1.1, dtype=tf.float32)
c = tf.constant(0.6, dtype=tf.float32)
print('***** PRIORS *****')  ## seems like a great way to initialize
print('a_prior: {}'.format(a_prior))
print('b_prior: {}'.format(b_prior))
print('c_prior: {}'.format(c_prior))
print('d_prior: {}'.format(d_prior))

## Model
I = tf.placeholder(tf.float32,[N,M])
zeta_u = Gamma(a_prior,a_prior/b_prior)
theta_u = Gamma(a, zeta_u)
eta_v = Gamma(c_prior, c_prior/d_prior)
beta_v = Gamma(c,eta_v)

U = Gamma(a*tf.ones([N,D]),tf.tile(tf.reshape(zeta_u, [-1, 1]), [N, D]))
V = Gamma(c*tf.ones([M,D]),tf.tile(tf.reshape(beta_v, [-1, 1]), [M, D]))
R = Gamma(tf.matmul(U,tf.transpose(V))*I,  tf.ones([N,M]))

print('***** PROBABILISTIC MODEL *****')
print('zeta_u: {}'.format(zeta_u))
print('beta_v: {}'.format(beta_v))
print('U: {}'.format(U))
print('V: {}'.format(V))
print('R: {}'.format(R))

## Inference
qU = Gamma(tf.Variable(tf.ones([N, D])), tf.nn.softplus(tf.Variable(tf.ones([N,D]))))
qV = Gamma(tf.Variable(tf.ones([M, D])), tf.nn.softplus(tf.Variable(tf.ones([M,D]))))
print('***** VARIATIONAL MODEL *****')
print('qU: {}'.format(qU))
print('qV: {}'.format(qV))

inference = ed.KLqp({U: qU,V: qV}, data={R :R_true, I: I_train})
inference.run(n_iter=10000, n_print=100)
```

**Comment by dustinvtran** [comment]
Poisson sampling is only available in a pre-release of TensorFlow; see https://discourse.edwardlib.org/t/poisson-distribution/70/2 for more discussion.

By NaN, do you mean it approaches NaN? In general, working with Gamma distributions with black box methods (as in KLqp) will typically fail. You'll need to use more model-specific methods.
